sugondat/chain/pallets/length-fee-adjustment/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+/// of an extrinsic to a fee.
+///
+/// By default, the implementation of `LengthToFee` is a constant multiplication. However, we
+/// aim to achieve a dynamic formula, thanks to the new multiplier, implementing `sp_weights::WeightToFee`
+/// for the Pallet struct and thus being able to use it as value for the associated type `LengthToFee` in the
+/// `pallet_transaction_type::Config`. The same pallet implements aslo `MultiplierUpdate` so it can be also
+/// used for the  `FeeMultiplierUpdate` associated type.
+///
+/// `targeted_length_fee_adjustment` is updated inside `on_finalize` while
+/// `targeted_weight_fee_adjustment` is updated by the method `Convert` called by the pallet `transaction_payment`.
+#[frame_support::pallet]
+pub mod pallet {
+
+    use frame_support::pallet_prelude::*;
+    use frame_system::pallet_prelude::*;
+    use pallet_transaction_payment::{
+        Multiplier, MultiplierUpdate, OnChargeTransaction, TargetedFeeAdjustment,
+    };
+    use sp_runtime::{
+        traits::{Convert, Get},
+        FixedPointNumber, Perquintill, SaturatedConversion, Saturating,
+    };
+
+    /// Configure the pallet by specifying the parameters and types on which it depends.
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_transaction_payment::Config {
+        // `targeted_weight_fee_adjustment` parameters
+        #[pallet::constant]
+        type TargetBlockFullness: Get<Perquintill>;
+        #[pallet::constant]
+        type AdjustmentVariableBlockFullness: Get<Multiplier>;
+        #[pallet::constant]
+        type MinimumMultiplierBlockFullness: Get<Multiplier>;
+        #[pallet::constant]
+        type MaximumMultiplierBlockFullness: Get<Multiplier>;
+
+        // `targeted_length_fee_adjustment` parameters
+        #[pallet::constant]
+        type TransactionByteFee: Get<<<Self as pallet_transaction_payment::Config>::OnChargeTransaction as OnChargeTransaction<Self>>::Balance>;
+        #[pallet::constant]
+        type MaximumBlockLength: Get<u32>;
+        #[pallet::constant]
+        type AdjustmentVariableBlockSize: Get<Multiplier>;
+        #[pallet::constant]
+        type MinimumMultiplierBlockSize: Get<Multiplier>;
+        #[pallet::constant]
+        type MaximumMultiplierBlockSize: Get<Multiplier>;
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    pub struct NextLengthMultiplierDefualt;
+    impl Get<Multiplier> for NextLengthMultiplierDefualt {
+        fn get() -> Multiplier {
+            Multiplier::saturating_from_integer(1)
+        }
+    }
+
+    #[pallet::storage]
+    pub type NextLengthMultiplier<T: Config> =
+        StorageValue<_, Multiplier, ValueQuery, NextLengthMultiplierDefualt>;
+
+    pub struct TargetBlockSizeDefualt;
+    impl Get<Perquintill> for TargetBlockSizeDefualt {
+        fn get() -> Perquintill {
+            Perquintill::from_percent(16) // 0.8MiB
```

Comment:
This is probably not the cleanest option (i.e. it's definitely not reusable for other codebases). Given the design constraints and downside, this seems good enough.

The "correct" way, as you point out, is likely something like a GenesisConfig or migration.

Reply:
yeah, I agree on what you just pointed out, maybe I could open an issue to replace this with some GenesisConfig logic and work on it as soon as all the other important features are there!